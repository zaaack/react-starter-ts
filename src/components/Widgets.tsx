// @flow
import React from 'react'
import styled from 'styled-components'
import { Route, Link, MemoryRouter, RouteComponentProps } from 'react-router-dom'
import { connect } from 'react-redux'
import { debug } from 'utils/index'
import Immuter from 'immuter'
import { clearFix } from 'polished'

export const StyledHome = styled.div`
  border: 1px solid #f90;
  padding: 0 10px;

  h2 {
    color: #9f0;
  }
`

function mapStateToProps(state) {
  return {
    home: state.home,
  }
}


export type HomeState = {
  title: string,
  errors: { [key: string]: string | boolean }
}
@connect(mapStateToProps)
@Immuter.bindComp()
class Home extends React.Component<RouteComponentProps<any>, HomeState> {
  state = {
    title: '',
    errors: {
      some: false,
    },
  }
  get: Function
  set: Function
  update: Function
  delete: Function
  unblock: Function

  componentDidMount() {
    const { props } = this
    debug('props', props)
    this.unblock = (props.history.block as any)((location, action) => {
      return '确定离开？'
    })
  }

  componentWillUnmount() {
    this.unblock()
    // for more API of immuter, you can see https://github.com/zaaack/immuter
    this.set('title', 'some')
    this.get('title')
    this.set('errors.some', 'Some errors happen!')
    this.set({
      'errors.some': 'Some errors happen!',
      'title': 'some',
    })
  }

  render() {
    const { props } = this
    ;(window as any).props = props
    return (
      <StyledHome>
        <h2>Home</h2>
      </StyledHome>
    )
  }
}

export { Home }

export const About = () => (
  <div>
    <h2>About2</h2>
  </div>
)

export const Topic = ({ match }: RouteComponentProps<any>) => (
  <div>
    <h3>
      {match.params.topicId}</h3>
  </div>
)

export const Topics = ({ match }: RouteComponentProps<any>) => (
  <div>
    <h2>Topics</h2>
    <ul>
      <li>
        <Link to={`${match.url}/rendering`}>
          Rendering with React
        </Link>
      </li>
      <li>
        <Link to={`${match.url}/components`}>
          Components
        </Link>
      </li>
      <li>
        <Link to={`${match.url}/props-v-state`}>
          Props v. State
        </Link>
      </li>
    </ul>

    <Route path={`${match.url}/:topicId`} component={Topic} />
    <Route exact path={match.url} render={() => (
      <h3>Please select a topic.</h3>
    )} />
  </div>
)
