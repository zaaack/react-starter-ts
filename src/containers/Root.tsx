// @flow
import * as React from 'react'
import { Provider } from 'react-redux'
import { withRouter } from 'react-router'
import { Route, Link } from 'react-router-dom'
import { Home, About, Topics } from '../components/Widgets'
import * as History from 'history'

import { ConnectedRouter } from 'react-router-redux'
import configureStore from '../store'

// Create a history of your choosing (we're using a browser history in this case)
let history: any = null

if (process.env.NODE_ENV === 'test') {
  const createMemoryHistory = require('history/createMemoryHistory').default
  history = createMemoryHistory()
  console.log('using createMemoryHistory for test')
} else {
  const createBrowserHistory = require('history/createBrowserHistory').default
  history = createBrowserHistory()
}

const store = configureStore(history)

export default function Root() {
  return (
    <Provider store={store}>
      { /* ConnectedRouter will use the store from Provider automatically */ }
      <ConnectedRouter history={history}>
        <div>
          <ul>
            <li><Link to="/">Home</Link></li>
            <li><Link to="/about">About</Link></li>
            <li><Link className="topics-link" to="/topics" onClick={e => {
              setTimeout(() => {
                console.log('Clicked!', history)
              }, 100)
            }}>Topics</Link></li>
          </ul>

          <hr />
          <Route exact path="/" component={Home} />
          <Route path="/about" component={About} />
          <Route path="/topics" component={Topics} />
        </div>
      </ConnectedRouter>
    </Provider>
  )
}

// Now you can dispatch navigation actions from anywhere!
// store.dispatch(push('/foo'))
