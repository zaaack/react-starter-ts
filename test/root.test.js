import test from 'ava'
import React from 'react'
import ReactDOM, { findDOMNode } from 'react-dom'
import render from 'react-test-renderer'
import Root from '../lib/containers/Root'
import { Home } from '../lib/components/Widgets'
import TestUtils from 'react-dom/test-utils'
import { shallow, mount } from 'enzyme'

async function renderIntoDocument(element) {
  const dom = document.createElement('div')
  // Fix instance is null bug in test-utils
  document.body.appendChild(dom)
  ReactDOM.render(element, dom)
  return dom
}

test('Root render', async t => {
  const tree = render.create(React.createElement(Root)).toJSON()
  t.snapshot(tree)
})

test('click routing', async t => {
  // JSDom doesn't support navigate, so React-Router should using createMemoryHistory
  // shallow cannot find children's children elements, so we use mount here
  const wrapper = mount(React.createElement(Root))
  // Don't know why jsdom doesn't contain evt.button=0,
  // but this break react-router-dom's Link
  // https://github.com/ReactTraining/react-router/blob/master/packages/react-router-dom/modules/Link.js#L41
  wrapper.find('a.topics-link').simulate('click', {button: 0})
  t.is(wrapper.find('h2').text(), 'Topics')
})
