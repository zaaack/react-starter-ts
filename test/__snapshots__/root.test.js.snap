// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Root render 1`] = `
<div>
  <ul>
    <li>
      <a
        href="/"
        onClick={[Function]}
      >
        Home
      </a>
    </li>
    <li>
      <a
        href="/about"
        onClick={[Function]}
      >
        About
      </a>
    </li>
    <li>
      <a
        className="topics-link"
        href="/topics"
        onClick={[Function]}
      >
        Topics
      </a>
    </li>
  </ul>
  <hr />
  <div
    className="bCYOQH"
  >
    <h2>
      Home
    </h2>
  </div>
</div>
`;

exports[`home render 1`] = `
<div
  className="bCYOQH"
>
  <h2>
    Home
  </h2>
</div>
`;
